### PR TITLE
Implement xml_serialize and xml_unserialize methods

### PR DESCRIPTION
**Pull request by jimhester**
This isn't completely done, it still needs tests, but wanted to send it and see if there were any API or implementation changes needed.

### CODE REVIEW

File: R/as_xml_document.R [hunk 1]
```diff
@@ -63,3 +63,17 @@ as_xml.list <- function(x, ..., missing = "elem") {
...
+}
+
+#' @export
+as_xml_document.xml_nodeset <- function(x, root, ...) {
+  doc <- xml_new_root(.value = root, ...)
+  for (i in seq_along(x)) {
+    xml_add_child(doc, x[[i]])
+  }
+  doc
+}
```
Comment (hadley) [inline R/as_xml_document.R]: Probably should add `as_xml_document.xml_document()` for completeness (should just return `x`)

File: R/xml_serialize.R [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+#'
+#' b <- xml_find_all(x, "//b")
+#' out <- xml_serialize(b, NULL)
+#' xml_unserialize(out)
+#' @export
+xml_serialize <- function(object, connection, ...) UseMethod("xml_serialize")
+
+#' @export
+xml_serialize.xml_document <- function(object, connection, ...) {
+  serialize(structure(as.character(object, ...), class = "xml_document"), connection)
```
Comment (hadley) [inline R/xml_serialize.R]: These should probably use unique class names like `xml_serialised_document` (just so if you load it another way you don't get confused)

File: R/xml_serialize.R [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+#' @export
+xml_serialize.xml_document <- function(object, connection, ...) {
```
Comment (hadley) [inline R/xml_serialize.R]: Check function name

File: R/xml_serialize.R [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+#' @rdname xml_serialize
+#' @export
+xml_unserialize <- function(connection, ...) {
+  object <- unserialize(connection)
+  if (inherits(object, "xml_nodeset")) {
+    x <- read_xml(unclass(object), ...)
+    return(xml_find_all(x, "/*/node()"))
+  }
```
Comment (hadley) [inline R/xml_serialize.R]: I'd say drop the `return()` and add `else`s

Comment (hadley) [inline R/xml_serialize.R]: Shouldn't the xpath just be `/node()` ? (if the intent is to handle nodesets with non-elements)

Comment (jimhester) [inline R/xml_serialize.R]: I just want to select all the direct children of the root, this was taken from http://stackoverflow.com/a/691787/2055486, possibly `/node()` would work...

File: R/xml_serialize.R (hunk 2 above)
Comment (jimhester) [inline R/xml_serialize.R]: Yeah that is a fair point, it is safer that way

File: R/xml_serialize.R [hunk 5]
```diff
@@ -17,33 +17,33 @@ xml_serialize <- function(object, connection, ...) UseMethod("xml_serialize")
...
+
+    # Select only the direct children of the root
+    xml_find_all(x, "/*/node()")
+  } else if (inherits(object, "xml_serialized_document")) {
     read_xml(unclass(object), ...)
-  }
-  if (inherits(object, "xml_node")) {
+  } else if (inherits(object, "xml_serialized_node")) {
     read_xml(unclass(object), ...)
   }
```
Comment (hadley) [inline R/xml_serialize.R]: Should also have an `else` here for safety

**Pull request closed without merging**